File: go/collatz-conjecture/collatz_conjecture.go
// Package collatzconjecture implements a method 
// to perform the Collatz Conjecture
package collatzconjecture

import "fmt"

// Function CollatzConjecture takes an int as
// input and returns the number of iterations it 
// takes to reach 1 from input.
// Returns an error if input is <= 0
func CollatzConjecture(input int) (int, error) {
	if input <= 0 {
		return -1, fmt.Errorf("Input must be positive - %v is non-positive", input)
	}

	numOfIterations := 0

	for input != 1 {

		if input % 2 == 0 {
			input /= 2
		} else {
			input = (3 * input) + 1
		}

		numOfIterations++
	}

	return numOfIterations, nil
}

File: cpp/collatz-conjecture/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include <stdexcept>
#include <iostream>

int collatz_conjecture::steps(long startingNumber)
{
	if (startingNumber < 1)
		throw std::domain_error("Starting number must be >= 1");

	if (startingNumber == 1)
		return 0;

	return startingNumber % 2 == 1 ? 2 + collatz_conjecture::steps((3 * startingNumber + 1) / 2) : 1 + collatz_conjecture::steps(startingNumber / 2);
}
